lib/spack/spack/concretize.py

```diff
@@ -347,26 +347,46 @@ def concretize_compiler_flags(self, spec):
             # running.
             return True
 
-        compiler_match = lambda other: (
-            spec.compiler == other.compiler and
-            spec.architecture == other.architecture)
+        def compiler_match(other, strict=True):
+            if strict:
+                # Other compiler already matches this compiler.
+                return (spec.compiler == other.compiler and
+                        spec.architecture == other.architecture)
+            else:
+                # Other compiler may match this compiler.
+                if other.compiler and other.compiler.concrete:
+                    return (spec.compiler == other.compiler and
+                            spec.architecture == other.architecture)
+                return True
 
         ret = False
         for flag in spack.spec.FlagMap.valid_compiler_flags():
             if flag not in spec.compiler_flags:
-                spec.compiler_flags[flag] = list()
-            try:
-                nearest = next(p for p in spec.traverse(direction='parents')
-                               if (compiler_match(p) and
-                                   (p is not spec) and
-                                   flag in p.compiler_flags))
-                nearest_flags = set(nearest.compiler_flags.get(flag, []))
-                flags = set(spec.compiler_flags.get(flag, []))
-                if (nearest_flags - flags):
-                    spec.compiler_flags[flag] = list(nearest_flags | flags)
-                    ret = True
-            except StopIteration:
-                pass
+                try:
+                    n = next(p for p in spec.traverse(direction='parents')
+                             if (compiler_match(p) and
+                                 (p is not spec) and
+                                 flag in p.compiler_flags))
+
+                    nearest_flags = set(n.compiler_flags[flag])
+                    flags = set()
+
+                    if (nearest_flags - flags) or n.compiler_flags[flag] == []:
```

Comment:
If it is true that ```flag in n.compiler_flags``` what are the possible values of ```n.compiler_flags[flag]```? I was assuming it would be a (possibly empty container of some sort). Can it be ```None```?